Scripts/Runtime/Jobs/Collections/DynamicBufferSharedWriteHandle.cs

```diff
@@ -0,0 +1,294 @@
+            m_SharedWriteSystems.Add(system);
+        }
+        
+        /// <summary>
+        /// Unregisters a <see cref="ComponentSystemBase"/> as a system that will shared write to
+        /// the <see cref="DynamicBuffer{T}"/>.
+        /// </summary>
+        /// <param name="system">The <see cref="ComponentSystemBase"/> that shared writes.</param>
+        public void UnregisterSystemForSharedWrite(ComponentSystemBase system)
+        {
+            Debug.Assert(system.World == m_World, $"System {system} is not part of the same world as this {nameof(DynamicBufferSharedWriteHandle<T>)}");
+            m_SharedWriteSystems.Remove(system);
+        }
+        
+        /// <summary>
+        /// Gets a <see cref="JobHandle"/> to be used to schedule the jobs that will shared writing to the
+        /// <see cref="DynamicBuffer{T}"/>.
+        /// </summary>
+        /// <param name="callingSystem">The <see cref="SystemBase"/> that is doing the shared writing.</param>
+        /// <param name="callingSystemDependency">The incoming Dependency <see cref="JobHandle"/> for the <paramref name="callingSystem"/></param>
+        /// <returns></returns>
+        public JobHandle GetSharedWriteJobHandle(SystemBase callingSystem, JobHandle callingSystemDependency)
+        {
+            Debug.Assert(m_SharedWriteSystems.Contains(callingSystem), $"Trying to get the shared write handle but {callingSystem} hasn't been registered. Did you call {nameof(RegisterSystemForSharedWrite)}?");
+            
+            //Rebuild our local cache if we need to. Will trigger a world cache rebuild if necessary too.
+            m_LocalCache.RebuildIfNeeded();
+            
+            //Ensure our local cache has the right order of systems that actually executed this frame
+            m_LocalCache.UpdateExecutedSystems(callingSystem);
+            
+            //Find out when our system executed in the order
+            int callingSystemOrder = m_LocalCache.GetExecutionOrderOf(callingSystem);
+
+            //If we're the first system to go in a frame, we're the first start point for shared writing.
+            if (callingSystemOrder == 0)
+            {
+                m_SharedWriteDependency = callingSystemDependency;
+            }
+            //Otherwise we want to check the system that executed before us to see what kind of lock they had 
+            //on our IBufferElementData
+            else
+            {
+                ComponentSystemBase previousSystem = m_LocalCache.GetSystemAtExecutionOrder(callingSystemOrder - 1);
+                
+                //If that system was a shared writable system, we don't want to move our dependency up so that we 
+                //can also share the write. If not, we move it up.
+                if (!m_SharedWriteSystems.Contains(previousSystem))
+                {
+                    m_SharedWriteDependency = callingSystemDependency;
+                }
+            }
```

Comment:
I don't think this is true.

`SystemA` executes but DOES NOT ask for the SharedWriteHandle because of the logic you described. 
`SystemB` executes and DOES ask for the SharedWriteHandle

`SystemB` will be delivered the incoming dependency of `SystemA` because that was when Shared Writing was allowed to start. `SystemA` just didn't use it but there's no reason `SystemB` couldn't start then. 

In a normal circumstance `SystemA` would have done a shared write too with it's own incoming dependency and then `SystemB` would have done a shared write with the same one. 

So no issues.

Reply:
Talked about this and there's a case here where:

- System1 SW
- System2 SW
- SystemE EW
- SystemA SW (But does not call GetSharedWriteDependency)
- SystemB SW - Possible error that the shared write dependency is pointing to System1 SW still.

Reply:
Added a fix for this.